### PR TITLE
Enable 'encoder' and 'tags_schema' highlighting settings at field level

### LINKED ISSUE #94028: Highlighting: `encoder` and `tags_schema` are not accepted on a field level
### Elasticsearch Version

8.6.2

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux fedora 6.1.11-200.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Feb  9 19:20:24 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Problem Description

When specifying either `encoder` or `tags_schema` configurations at a field level in a highlighting query leads to an exception in the response:
```json
"root_cause": [
      {
        "type": "x_content_parse_exception",
        "reason": "[12:9] [highlight_field] unknown field [encoder]"
      }
    ],
```
Also, looking at the documentation - https://www.elastic.co/guide/en/elasticsearch/reference/8.6/highlighting.html doesn't specify any of the settings being as global only and the override section tells that any settings can be overridden at a field level.

### Steps to Reproduce

1. Create an index with an analyzed property:
```http
PUT http://localhost:9200/highlight-test-000001
Content-Type: application/json

{
  "mappings": {
    "dynamic": "strict",
    "properties": {
      "string": {
        "type": "text",
        "analyzer": "custom-analyzer"
      }
    }
  },
  "settings": {
    "analysis": {
      "analyzer": {
        "custom-analyzer": {
          "type": "standard",
          "stopwords": "_english_"
        }
      }
    }
  }
}
```
2. Add some documents:
```http
PUT http://localhost:9200/highlight-test-000001/_doc/1
Content-Type: application/json

{
  "string": "some text goes here"
}
```
```http
PUT http://localhost:9200/highlight-test-000001/_doc/2
Content-Type: application/json

{
  "string": "<!DOCTYPE html><html><head><title>Page Title</title></head><body><h1>This is a Heading</h1><p>This is a text.</p></body></html>"
}
```
3. Try to execute a search query with highlighting and specifying settings on a field level:
```http
POST http://localhost:9200/highlight-test-000001/_search
Content-Type: application/json

{
  "query": {
    "match": {
      "string": {
        "query": "text"
      }
    }
  },
  "highlight": {
    "fields": {
      "string": {
        "encoder": "html"
      }
    }
  }
}
```

leads to:
```json
{
  "error": {
    "root_cause": [
      {
        "type": "x_content_parse_exception",
        "reason": "[12:9] [highlight_field] unknown field [encoder]"
      }
    ],
    "type": "x_content_parse_exception",
    "reason": "[12:20] [highlight] failed to parse field [fields]",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": "[12:20] [fields] failed to parse field [string]",
      "caused_by": {
        "type": "x_content_parse_exception",
        "reason": "[12:9] [highlight_field] unknown field [encoder]"
      }
    }
  },
  "status": 400
}
```

Similarly using the `tags_schema` :
```http
POST http://localhost:9200/highlight-test-000001/_search
Content-Type: application/json

{
  "query": {
    "match": {
      "string": {
        "query": "text"
      }
    }
  },
  "highlight": {
    "fields": {
      "string": {
        "tags_schema": "styled"
      }
    }
  }
}
```
leads to the same error:
```json
{
  "error": {
    "root_cause": [
      {
        "type": "x_content_parse_exception",
        "reason": "[12:9] [highlight_field] unknown field [tags_schema]"
      }
    ],
    "type": "x_content_parse_exception",
    "reason": "[12:24] [highlight] failed to parse field [fields]",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": "[12:24] [fields] failed to parse field [string]",
      "caused_by": {
        "type": "x_content_parse_exception",
        "reason": "[12:9] [highlight_field] unknown field [tags_schema]"
      }
    }
  },
  "status": 400
}
```

Specifying any of these settings at a global level works fine. With for example a query like:
```http
POST http://localhost:9200/highlight-test-000001/_search
Content-Type: application/json

{
  "query": {
    "match": {
      "string": {
        "query": "text"
      }
    }
  },
  "highlight": {
    "fields": {
      "string": {}
    },
    "tags_schema": "styled",
    "encoder": "html"
  }
}
```
the result on the sample data set is:
```json
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 0.24649063,
    "hits": [
      {
        "_index": "highlight-test-000001",
        "_id": "1",
        "_score": 0.24649063,
        "_source": {
          "string": "some text goes here"
        },
        "highlight": {
          "string": [
            "some <em class=\"hlt1\">text</em> goes here"
          ]
        }
      },
      {
        "_index": "highlight-test-000001",
        "_id": "2",
        "_score": 0.1446617,
        "_source": {
          "string": "<!DOCTYPE html><html><head><title>Page Title</title></head><body><h1>This is a Heading</h1><p>This is a text.</p></body></html>"
        },
        "highlight": {
          "string": [
            "DOCTYPE html&gt;&lt;html&gt;&lt;head&gt;&lt;title&gt;Page Title&lt;&#x2F;title&gt;&lt;&#x2F;head&gt;&lt;body&gt;&lt;h1&gt;This is a Heading&lt;&#x2F;h1&gt;&lt;p&gt;This is a <em class=\"hlt1\">text</em>"
          ]
        }
      }
    ]
  }
}
```

Elasticsearch is running in docker using:
```bash
docker run --rm -ti -p 9200:9200 -p 9300:9300 --ulimit memlock=-1:-1 --memory-swappiness=0  --name es-8.6-it -e "discovery.type=single-node" -e "xpack.security.enabled=false" -e "ES_JAVA_OPTS=-Xms1g -Xmx1g" -e "cluster.routing.allocation.disk.threshold_enabled=false" elastic/elasticsearch:8.6.2

```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by Leuteris**
Executing a search query with highlighting and specifying `encoder` or `tags_schema` settings at the field level was resulting in an x_content_parse_exception.

This PR moves `encoder` and `tags_schema` highlighting setting on `AbstractHighlighterBuilder`, allowing them to be set global and at field level. 

Closes #94028

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elasticmachine generate changelog

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java [hunk 1]
```diff
@@ -173,6 +178,7 @@ public final void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalStringArray(postTags);
         out.writeOptionalVInt(fragmentSize);
         out.writeOptionalVInt(numOfFragments);
+        out.writeOptionalString(encoder);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:181]: Similar code about the inter-node communication as above.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java [hunk 2]
```diff
@@ -124,15 +120,13 @@ public HighlightBuilder(HighlightBuilder template, QueryBuilder highlightQuery,
      */
     public HighlightBuilder(StreamInput in) throws IOException {
         super(in);
-        encoder(in.readOptionalString());
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java:127]: Similar code about the inter-node communication as above.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java [hunk 3]
```diff
@@ -124,15 +120,13 @@ public HighlightBuilder(HighlightBuilder template, QueryBuilder highlightQuery,
...
         useExplicitFieldOrder(in.readBoolean());
         this.fields = in.readCollectionAsList(Field::new);
         assert this.equals(new HighlightBuilder(this, highlightQuery, fields)) : "copy constructor is broken";
     }
 
     @Override
     protected void doWriteTo(StreamOutput out) throws IOException {
-        out.writeOptionalString(encoder);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java:135]: Similar code about the inter-node communication as above.

**Review by mayya-sharipova — COMMENTED** [review]
@Leuteris Thanks for your contribution. Overall, the PR looks good to me. Let's fix how the over-wire format of Highlighters.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java [hunk 4]
```diff
@@ -137,6 +141,7 @@ protected AbstractHighlighterBuilder(StreamInput in) throws IOException {
         postTags(in.readOptionalStringArray());
         fragmentSize(in.readOptionalVInt());
         numOfFragments(in.readOptionalVInt());
+        encoder(in.readOptionalString());
```
[the remark this answers is not in the document]
Comment (Leuteris) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java]: Thank you for your detailed feedback. The transport version has been added.

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilderTests.java [hunk 5]
```diff
@@ -477,6 +476,49 @@ public void testParsingTagsSchema() throws IOException {
...
+            assertArrayEquals(
+                "setting tags_schema 'default' at the field level should alter post_tags for that that field",
+                HighlightBuilder.DEFAULT_POST_TAGS,
+                defaultStringField.postTags()
+            );
+
+            Field styledStringField = getFieldBuilderByName(highlightBuilder, "styled_string");
+            assertNotNull(styledStringField);
+            assertArrayEquals(
+                "setting tags_schema 'default' at the field level should alter pre_tags for that that field",
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilderTests.java:512]: nit: "setting tags_schema 'default" -> "setting tags_schema 'styled'

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine test this please

**Review by mayya-sharipova — APPROVED** [review]
@Leuteris Thanks for a great work! This LGTM!

I've left a small comment, please address it.

For merging, we need to merge the recent changes from main branch to this PR, and update your changes in `TransportVersions.java` to increase `HIGHLIGHTERS_TAGS_ON_FIELD_LEVEL` version to the next available.

**Comment by mayya-sharipova** [comment]
@Leuteris Looks like it needs another update with the main branch.  Sorry `TransportVersions.java` file changes very fast.

**Comment by mayya-sharipova** [comment]
@Leuteris Thanks for your great work. Merged for 8.14.

**Pull request merged**